src/activity_actor.cpp

```diff
@@ -2204,6 +2206,196 @@ std::unique_ptr<activity_actor> stash_activity_actor::deserialize( JsonIn &jsin
     return actor.clone();
 }
 
+bool reload_activity_actor::can_reload() const
+{
+    if( reload_targets.size() != 2 || quantity <= 0 ) {
+        debugmsg( "invalid arguments to ACT_RELOAD" );
+        return false;
+    }
+
+    if( !reload_targets[0] ) {
+        debugmsg( "reload target is null, failed to reload" );
+        return false;
+    }
+
+    if( !reload_targets[1] ) {
+        debugmsg( "ammo target is null, failed to reload" );
+        return false;
+    }
+
+    return true;
+}
+
+void reload_activity_actor::start( player_activity &act, Character &/*who*/ )
+{
+    act.moves_total = moves_total;
+    act.moves_left = moves_total;
+}
+
+void reload_activity_actor::reload_gun( Character &who, item &reloadable, item &ammo ) const
+{
+    const bool ammo_uses_speedloader = ammo.has_flag( flag_SPEEDLOADER );
+    who.recoil = MAX_RECOIL;
+    if( reloadable.has_flag( flag_RELOAD_ONE ) && !ammo_uses_speedloader ) {
+        for( int i = 0; i != quantity; ++i ) {
+            add_msg( m_neutral, _( "You insert one %2$s into the %1$s." ), reloadable.tname(), ammo.tname() );
+        }
+    }
+    make_reload_sound( who, reloadable );
+}
+
+void reload_activity_actor::make_reload_sound( Character &who, item &reloadable ) const
+{
+    if( reloadable.type->gun->reload_noise_volume > 0 ) {
+        sfx::play_variant_sound( "reload", reloadable.typeId().str(),
+                                 sfx::get_heard_volume( who.pos() ) );
+        sounds::ambient_sound( who.pos(), reloadable.type->gun->reload_noise_volume,
+                               sounds::sound_t::activity, reloadable.type->gun->reload_noise.translated() );
+    }
+}
+
+void reload_activity_actor::finish( player_activity &act, Character &who )
+{
+    act.set_to_null();
+    if( !can_reload() ) {
+        return;
+    }
+
+    item &reloadable = *reload_targets[ 0 ];
+    item &ammo = *reload_targets[ 1 ];
+    std::string reloadable_name = reloadable.tname();
+    std::string ammo_name = ammo.tname();
+    // cache check results because reloading deletes the ammo item
+    const bool ammo_is_filthy = ammo.is_filthy();
+
+    if( !reloadable.reload( who, std::move( reload_targets[ 1 ] ), quantity ) ) {
```

Comment:
I think this `std::move` invalidates the `ammo` const ref defined above and that's why ASAN errors out

Reply:
`std::move` doesn't really "move" the data, it's just makes an r value from an l value. You are not zeroing the `reload_targets[ 1 ] `

Reply:
I read something else in the std::move [documentation](https://en.cppreference.com/w/cpp/utility/move). Either way, `item::reload` is deleting the item in `item.cpp:8247`:
```
        ammo.remove_item();
```
even if the `item_location` `ammo` is still somehow valid after the `std::move`.

We've had this exact same issue [before](https://github.com/CleverRaven/Cataclysm-DDA/issues/42089#issuecomment-658717453). Take a look at the ASan error [too](https://github.com/CleverRaven/Cataclysm-DDA/runs/1376595179#step:6:580).

Reply:
>  ammo.remove_item();

That I can't argue with. The solution would be to just remove ammo as a reference and make a copy of it so it would be 
`item ammo = *reload_targets[ 1 ];` I will issue a new PR in spare time. Thanks.

As for docs, I don't really see where you read that it's invalidating the previous object/moving to a new address.

Reply:
Looking at #45545 I've got the reference to this PR and I would argue with both of you. @J-Cieplinski is correct, `std::move` itself doesn't do the magic and just make r-value reference from l-value, but that may trigger destructable move semantic inside any underlying object. So if you don't have any actual "move" functionality for the object, it doesn't make sense to move it, because it's the same as copying it. But if you do have implemented move functionality, than the rule of thump is to keep object in correct only for destruction state, that's it, it is not supposed to be used for something else.
And exact at this line you are moving object that is holding `std::shared_ptr` and it does have fully-functional move semantic, after this point your `item &ammo` reference is invalid and it's just a matter of time to get segfault.

Reply:
>  `std::shared_ptr` and it does have fully-functional move semantic

Just a minor nitpick here: the move constructor for `std::shared_ptr` is fully specified and it leaves the original ptr empty (as you can see in number 10 [here](https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr). You're right about the rest and thanks for clarifying.